### ISSUE TITLE
-Wmissing-field-initializers warning missing for field designators

### ISSUE THREAD

**Issue by PiJoules** [llvm/llvm-project#42576]
|  |  |
| --- | --- |
| Bugzilla Link | [43231](https://llvm.org/bz43231) |
| Version | unspecified |
| OS | All |
| CC | @DougGregor,@zygoloid,@tamird |

## Extended Description 
Given the following code

```
// ~/misc/test.cpp
struct foo {
  int x;
  int y;
};

void func() {
  struct foo f = { 11 };
  (void) f;
}
```

And compiling with `bin/clang++ ~/misc/test.cpp -c -Wmissing-field-initializers`, we get the missing field initializers warning.

```
$ bin/clang++ ~/misc/test.cpp -c -Wmissing-field-initializers
/usr/local/google/home/leonardchan/misc/test.cpp:7:23: warning: missing field 'y' initializer [-Wmissing-field-initializers]
  struct foo f = { 11 };
                      ^
1 warning generated.
```

But if we instead initialize `f` with `{ .x = 11 }`, we get no warning. In both cases, y is still uninitialized but we only get the warning for the first case. I think we should emit this diagnostic in both cases.

**Comment by wheatman** [comment]
This was fixed between 16 and trunk(68e94f1f27b658183c3ebbce4aaadc437d0cee3a)
https://godbolt.org/z/zoTYKbYdc

code
```
// ~/misc/test.cpp
struct foo {
  int x;
  int y;
};

void func() {
  struct foo f = { 11 };
  struct foo g = { .x = 11 };
  (void) f;
  (void) g;
}
```

old warning
```
<source>:8:23: warning: missing field 'y' initializer [-Wmissing-field-initializers]
  struct foo f = { 11 };
                      ^
1 warning generated.
Compiler returned: 0
```

new warning
```
<source>:8:23: warning: missing field 'y' initializer [-Wmissing-field-initializers]
    8 |   struct foo f = { 11 };
      |                       ^
<source>:9:28: warning: missing field 'y' initializer [-Wmissing-field-initializers]
    9 |   struct foo g = { .x = 11 };
      |                            ^
2 warnings generated.
Compiler returned: 0```
